File: src/main/java/com/example/dpop/orchestrator/session/ProcessSessionRepository.java
package com.example.dpop.orchestrator.session;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Repository
public interface ProcessSessionRepository extends JpaRepository<ProcessSession, UUID> {
    Optional<ProcessSession> findByChannelSessionIdAndPurpose(UUID channelSessionId, ProcessPurpose purpose);

    List<ProcessSession> findByChannelSessionIdOrderByCreatedAtDesc(UUID channelSessionId);
}

File: src/main/kotlin/com/example/dpop/orchestrator/session/ProcessSessionRepository.kt
package com.example.dpop.orchestrator.session

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository
import java.time.Instant
import java.util.UUID

@Repository
interface ProcessSessionRepository : JpaRepository<ProcessSession, UUID> {
    /** At most one active process per channel (docs/07-betrieb.md #2). */
    fun findFirstByChannelSessionIdAndStateOrderByCreatedAtDesc(
        channelSessionId: UUID,
        state: ProcessState
    ): ProcessSession?

    /** Retention clock starts at consumedAt or expiresAt, whichever applies (docs/07-betrieb.md #3). */
    fun deleteByConsumedAtBeforeOrExpiresAtBefore(consumedCutoff: Instant, expiresCutoff: Instant): Long
}
